src/main/scala/is/hail/variant/GATKGenotypeView.scala

```diff
@@ -0,0 +1,265 @@
+package is.hail.variant
+
+import is.hail.annotations.{MemoryBuffer, RegionValue, UnsafeRow, UnsafeUtils}
+import is.hail.expr._
+import is.hail.utils._
+
+object GATKGenotypeView {
+  def apply(rowSignature: TStruct): GATKGenotypeView = {
+    rowSignature.fields(2).typ.asInstanceOf[TArray].elementType match {
+      case TGenotype => new TGenotypeView(rowSignature)
+      case _: TStruct => new StructGenotypeView(rowSignature)
+      case t => fatal(s"invalid genotype representation: $t, expect TGenotype or TStruct")
+    }
+  }
+}
+
+// FIXME: This is removed, and StructGenotypeView becomes the only genotype view when TGenotype is removed
+sealed abstract class GATKGenotypeView {
+  def setRegion(mb: MemoryBuffer, offset: Int)
+
+  def setGenotype(idx: Int)
+
+  def hasGT: Boolean
+
+  def hasAD: Boolean
+
+  def hasDP: Boolean
+
+  def hasGQ: Boolean
+
+  def hasPL: Boolean
+
+  def getGT: Int
+
+  def getAD(idx: Int): Int
+
+  def getDP: Int
+
+  def getGQ: Int
+
+  def getPL(idx: Int): Int
+}
+
+private class TGenotypeView(rs: TStruct) extends GATKGenotypeView {
+  private val ta = rs.fields(2).typ.asInstanceOf[TArray]
+  private val t = TGenotype.representation
+  private val tSize = ta.elementByteSize
+  private val gsOffset = rs.byteOffsets(2)
+
+  private val gtIndex = t.fieldIdx("gt")
+  private val gtOffset = t.byteOffsets(gtIndex)
+
+  private val adIndex = t.fieldIdx("ad")
+  private val adOffset = t.byteOffsets(adIndex)
+
+  private val dpIndex = t.fieldIdx("dp")
+  private val dpOffset = t.byteOffsets(dpIndex)
+
+  private val gqIndex = t.fieldIdx("gq")
+  private val gqOffset = t.byteOffsets(gqIndex)
+
+  private val pxIndex = t.fieldIdx("px")
+  private val pxOffset = t.byteOffsets(pxIndex)
+
+  private val linearScaleIndex = t.fieldIdx("isLinearScale")
+  private val linearScaleOffset = t.byteOffsets(linearScaleIndex)
+
+  private var m: MemoryBuffer = _
+  private var arrStart: Int = _
+  private var arrSize: Int = _
+  private var elementsStart: Int = _
+  private var currentElementStart: Int = _
+  private var elementIsNull: Boolean = _
```

Comment:
I'd call these `gs...` or `g...`, e.g. `gsStart`, `gIsNull`.

We seem to be using the term missing instead of null.

I've been using off(set) instead of start.

I'd like to keep naming consistent.  We can change my names if you prefer, but we should keep them consistent.